google/cloud/pubsublite/internal/stream_factory.h

```diff
@@ -45,74 +52,111 @@ inline std::unique_ptr<grpc::ClientContext> MakeGrpcClientContext(
   return context;
 }
 
+template <class Request, class Response>
+StreamFactory<Request, Response> AddClientMetadata(
+    ClientMetadata const& metadata,
+    ContextStreamFactory<Request, Response> underlying) {
+  return [=] { return underlying(MakeGrpcClientContext(metadata)); };
+}
+
+template <typename Request, typename Response>
+inline ContextStreamFactory<Request, Response> AddAuthContext(
+    std::shared_ptr<google::cloud::internal::GrpcAuthenticationStrategy> const&
+        auth,
+    ContextStreamFactory<Request, Response> const& underlying) {
+  return [=](std::unique_ptr<grpc::ClientContext> context) {
+    return absl::make_unique<
+        internal::AsyncStreamingReadWriteRpcAuth<Request, Response>>(
+        std::move(context), auth,
+        [=](std::unique_ptr<grpc::ClientContext> ctx) {
+          return underlying(std::move(ctx));
+        });
+  };
+}
+
+template <class Obj, class R, class... Args>
```

Comment:
I would have just written:

```cc
template <class Functor, class Request, class Response>
inline ContextStreamFactory<Request, Response> MakeRawStreamFactory(
    google::cloud::CompletionQueue const& cq, Function fun) {
  return [=](std::unique_ptr<grpc::ClientContext> context)
             -> std::unique_ptr<BidiStream<Request, Response>> {
    return internal::MakeStreamingReadWriteRpc<Request, Response>(
        cq, std::move(context), fun);
  };
}
```

and pass a lambda as `fun`, for example:

```cc
inline StreamFactory<pubsublite::v1::SubscribeRequest,
                     pubsublite::v1::SubscribeResponse>
MakeStreamFactory(
    std::shared_ptr<pubsublite::v1::SubscriberService::StubInterface> const&
        stub,
    google::cloud::CompletionQueue const& cq,
    std::shared_ptr<google::cloud::internal::GrpcAuthenticationStrategy> const&
        auth,
    ClientMetadata const& metadata = {}) {
  return AddClientMetadata(
      metadata,
      AddAuthContext(auth, MakeRawStreamFactory(
                               cq, [stub](grpc::ClientContext* ctx, grpc::CompletionQueue* cq) {
          return stub->PrepareAsyncSubscribe(ctx, cq); }
   )));
}
```

Reply:
The way you've written it, MakeRawStreamFactory needs to provide the request and response arguments at the call site, i.e. `MakeRawStreamFactory<pubsublite::v1::SubscribeRequest, pubsublite::v1::SubscribeResponse>` since there's no where but the functor where the types can be deduced. I can make this change, but I think the current structure is more clear.